src/main/scala/com/nec/vectorengine/VeProcessImpl.scala

```diff
@@ -92,6 +96,29 @@ final case class WrappingVeo private (val node: Int,
     }
   }
 
+  private[vectorengine] def withVeoThread[T](thunk: veo_thr_ctxt => T): T = {
+    // get context with (approximately) shortest queue
+    val (lock, ctx) = contextLocks.minBy { case (lock, _) => lock.getQueueLength }
+    withVeoThreadLock(lock) { thunk(ctx) }
+  }
+
+  private[vectorengine] def withVeoThread[T](ctx: veo_thr_ctxt)(thunk: veo_thr_ctxt => T): T = {
+    // find lock for specified context
+    val (lock, _) = contextLocks.find{case (_, c) => ctx == c}.get
+    withVeoThreadLock(lock) { thunk(ctx) }
+  }
+
+  private[vectorengine] def withVeoThreadLock[T](lock: ReentrantReadWriteLock)(thunk: => T): T = {
+    lock.writeLock().lock()
+    try {
+      withVeoProc{
```

Comment:
I think we should have `withVeoProc` first before `withVeoThread`, bc `withVeoProc` checks whether the process is open or not.

```
withVeoProc {
  withVeoThread { ctx =>
    thunk(ctx)
  }
}
```

That way we can short circuit before attempting to fetch the lock associated with one of the thread contexts